File: LL.java
class LL {
	
	class Node{
		int data;
		Node next;
	
	Node(int data){
		this.data = data;
		this.next = null;
		}
	}
	Node head;
	
	public void addFirst(int data){
		Node newNode = new Node(data);
		
		if(head == null)
		{
				head = newNode;
				return;
		}
		newNode.next = head;
		head = newNode;
	}
	
	public void addLast(int data){
		Node newNode = new Node(data);
		if(head == null){
			head = newNode;
			return;
		}
		
		Node curr = head;
		while(curr.next != null){
			curr = curr.next;
		}
		curr.next = newNode;
	}
	
	
	public void printList(){
		if(head == null){
			System.out.println("empty");
			return;
		}
		Node curr = head;
		while(curr != null){
			System.out.print(curr.data + " ");
			curr = curr.next;
		}
		
	}
	
	public void deleteFirst(){
		if(head == null){
			System.out.println("empty");
			return;
		}
		head = head.next;
	}
	
	public void deleteLast(){
	if(head == null){
			System.out.println("empty");
			return;
		}
	if(head.next == null){
		head = null;
		return;
	}
		Node secondLast = head;
		Node lastNode = head.next;
		while(lastNode.next != null){
			lastNode = lastNode.next;
			secondLast = secondLast.next;
			
		}
		secondLast.next = null;
	}
	
	public void deleteAtIndex(int index){
          Node now=head;
          Node prev=null;
          int j=1;
          while(j< index){
              prev = now;
	       	  now = now.next;
		      j++;
	             	}
		            prev.next = now.next;
		            now.next = null;
	}
	
	public void addAtIndex(int data,int index){
	    Node ptr=new Node(data);
	    Node p=head;
	    int i=0;
        while(i!=index-1)
        {
          p=p.next;
          i++;
        }
        ptr.data=data;
        ptr.next=p.next;
        p.next=ptr;
	}
	
	

	
	public static void main(String args[]){
		LL list = new LL();
	    list.addFirst(1);
	    list.addAtIndex(3,1);
	    list.printList();
	    System.out.println();
	    list.addFirst(5);
	    list.printList();
	    System.out.println();
	    list.deleteLast();
	    list.printList();
	    
	    
	    
		
	}
}

File: LL.c
#include <stdio.h>
#include<stdlib.h>
struct Node{
    int data;
    struct Node *next;
};
void printList(struct Node *ptr)
{
    while(ptr!=NULL)
    {
        
        printf("%d ",ptr->data);
        ptr=ptr->next;
    }
    printf("\n");
}
struct Node *insertAtIndex(struct Node *head,int data,int index)
{
    struct Node *ptr=(struct Node*)malloc(sizeof(struct Node));
    struct Node *p=head;
    int i=0;
    while(i!=index-1)
    {
        p=p->next;
        i++;
    }
    ptr->data=data;
    ptr->next=p->next;
    p->next=ptr;
    return head;
}

struct Node *deleteAtIndex(struct Node *head,int index)
{
    struct Node *p=head;
    struct Node *q=head->next;
    for(int i=0;i<index-1;i++)
    {
        p=p->next;
        q=q->next;
    }
    p->next=q->next;
    free(q);
    return head;
}

struct Node * insertAtEnd(struct Node *head, int data){
    struct Node * ptr = (struct Node *) malloc(sizeof(struct Node));
    ptr->data = data;
    struct Node * p = head;
 
    while(p->next!=NULL){
        p = p->next;
    }
    p->next = ptr;
    ptr->next = NULL;
    return head;
}

struct Node * insertAtFirst(struct Node *head, int data){
    struct Node * ptr = (struct Node *) malloc(sizeof(struct Node));
    ptr->data = data;
    ptr->next = head;
    return ptr; 
}

struct Node *deleteAtFirst(struct Node *head){
    head=head->next;
    return head;
}

struct Node *deleteAtLast(struct Node *head){
    struct Node * p =head;
    struct Node *q=head->next;
    while(q->next!=NULL)
    {
        p=q;
        q=q->next;
    }
    p->next=NULL;
}




    
int main()
{
    struct Node *head;
    struct Node *second;
    struct Node *third;
    head=(struct Node *)malloc(sizeof(struct Node));
    second=(struct Node *)malloc(sizeof(struct Node));
    third=(struct Node *)malloc(sizeof(struct Node));
    head->data=10;
    head->next=second;
    second->data=20;
    second->next=third;
    third->data=30;
    third->next=NULL;
    printList(head);
    insertAtIndex(head,40,1);
    printList(head);
    deleteAtLast(head);
    head=insertAtFirst(head,4);
    printList(head);

    return 0;
}